Fix tablemergekeys to behave like documentation

fmassa opened the pull request:
Hi,

So, I'm not 100% sure about this PR, but *as is* the documentation doesn't match the behaviour of `transforms.tablemergekeys`.
Indeed, it currently iterates via `ipairs`, outputting an empty table for the example in the doc.

I suppose that the `pairs` iterators was there before, any idea why it was changed to `ipairs`?

In the doc I also mentioned (implicitly) that it returns a function instead of the merged table.

Sorry about the noise in the documentation, the extra lines should correspond to the missing auto-documentation generation from https://github.com/torchnet/torchnet/pull/15 and https://github.com/torchnet/torchnet/pull/17 .

**PS**: `th -largcheck.dump -e "require 'torchnet'"` doesn't work for the documentation generation, I had to use `luajit -largcheck.dump -e "require 'torchnet'"` instead.

lvdmaaten commented:
I don't think you're supposed to use this function with non-numeric "keys". The documentation refers to `sample1`, `sample2`, etc. for clarity of presentation, but the internal representation would really just have numeric keys.

I don't think we want to replace `ipairs` by `pairs` here: `pairs` does not give any guarantees on the order in which elements are processed, which may lead to bugs or unexpected behavior.

fmassa commented:
I see, this lack of guarantees on the iteration order of `ipairs` is one of the reasons that was making me hesitate about this PR.

But giving the way `tablemergekeys` is implemented, I think the lack of order wouldn't be a problem, except if one feeds to it a vector with a hole in it (previous implementation would stop in the hole, whereas this one would simply go over all elements).

One thing that might be nice though about the `ipairs` is that it would make the transformation an involution, meaning `f(f(x)) = x`.
Also, the use-case that brought me to this PR was while creating a toy dataset,
```lua
ds = {input=torch.rand(10,2):split(1,1), labels=torch.rand(10):split(1,1)}
```
and I thought about using `tablemergekeys` to easily process it to feed into a `TableDataset`.

Anyway, it might be indeed a risky change. I'll stop chatting and I'll close the PR :-)
